File: bindings/go/satox/quantum/post_quantum_algorithms_test.go
package quantum

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestNewPostQuantumAlgorithms(t *testing.T) {
	pqa := NewPostQuantumAlgorithms()
	assert.NotNil(t, pqa)
	assert.False(t, pqa.IsInitialized())
}

func TestInitialize(t *testing.T) {
	pqa := NewPostQuantumAlgorithms()
	err := pqa.Initialize()
	assert.NoError(t, err)
	assert.True(t, pqa.IsInitialized())

	// Test double initialization
	err = pqa.Initialize()
	assert.NoError(t, err)
	assert.True(t, pqa.IsInitialized())
}

func TestShutdown(t *testing.T) {
	pqa := NewPostQuantumAlgorithms()
	err := pqa.Initialize()
	assert.NoError(t, err)

	err = pqa.Shutdown()
	assert.NoError(t, err)
	assert.False(t, pqa.IsInitialized())

	// Test shutdown when not initialized
	err = pqa.Shutdown()
	assert.NoError(t, err)
}

func TestGetAvailableAlgorithms(t *testing.T) {
	pqa := NewPostQuantumAlgorithms()
	err := pqa.Initialize()
	assert.NoError(t, err)

	algorithms, err := pqa.GetAvailableAlgorithms()
	assert.NoError(t, err)
	assert.NotEmpty(t, algorithms)

	// Verify known algorithms are present
	knownAlgorithms := []string{"CRYSTALS-Kyber", "NTRU", "SABER", "CRYSTALS-Dilithium", "FALCON", "SPHINCS+"}
	for _, algo := range knownAlgorithms {
		assert.Contains(t, algorithms, algo)
	}
}

func TestGetAlgorithmInfo(t *testing.T) {
	pqa := NewPostQuantumAlgorithms()
	err := pqa.Initialize()
	assert.NoError(t, err)

	// Test with valid algorithm
	info, err := pqa.GetAlgorithmInfo("CRYSTALS-Kyber")
	assert.NoError(t, err)
	assert.Equal(t, "CRYSTALS-Kyber", info.Name)
	assert.Greater(t, info.SecurityLevel, 0)
	assert.Greater(t, info.KeySize, 0)
	assert.Greater(t, info.SignatureSize, 0)
	assert.NotEmpty(t, info.Description)

	// Test with invalid algorithm
	_, err = pqa.GetAlgorithmInfo("InvalidAlgorithm")
	assert.Error(t, err)
}

func TestIsAlgorithmAvailable(t *testing.T) {
	pqa := NewPostQuantumAlgorithms()
	err := pqa.Initialize()
	assert.NoError(t, err)

	// Test with valid algorithm
	available, err := pqa.IsAlgorithmAvailable("CRYSTALS-Kyber")
	assert.NoError(t, err)
	assert.True(t, available)

	// Test with invalid algorithm
	available, err = pqa.IsAlgorithmAvailable("InvalidAlgorithm")
	assert.NoError(t, err)
	assert.False(t, available)
}

func TestIsAlgorithmRecommended(t *testing.T) {
	pqa := NewPostQuantumAlgorithms()
	err := pqa.Initialize()
	assert.NoError(t, err)

	// Test with recommended algorithm
	recommended, err := pqa.IsAlgorithmRecommended("CRYSTALS-Kyber")
	assert.NoError(t, err)
	assert.True(t, recommended)

	// Test with non-recommended algorithm
	recommended, err = pqa.IsAlgorithmRecommended("InvalidAlgorithm")
	assert.NoError(t, err)
	assert.False(t, recommended)
}

func TestGenerateKeyPairForAlgorithm(t *testing.T) {
	pqa := NewPostQuantumAlgorithms()
	err := pqa.Initialize()
	assert.NoError(t, err)

	// Test with valid algorithm
	publicKey, privateKey, err := pqa.GenerateKeyPair("CRYSTALS-Kyber")
	assert.NoError(t, err)
	assert.NotNil(t, publicKey)
	assert.NotNil(t, privateKey)

	// Test with invalid algorithm
	_, _, err = pqa.GenerateKeyPair("InvalidAlgorithm")
	assert.Error(t, err)
}

func TestSignAndVerify(t *testing.T) {
	pqa := NewPostQuantumAlgorithms()
	err := pqa.Initialize()
	assert.NoError(t, err)

	// Generate key pair
	publicKey, privateKey, err := pqa.GenerateKeyPair("CRYSTALS-Dilithium")
	assert.NoError(t, err)

	// Test data
	data := []byte("Test message for signing")

	// Sign data
	signature, err := pqa.Sign(privateKey, data)
	assert.NoError(t, err)
	assert.NotNil(t, signature)

	// Verify signature
	valid, err := pqa.Verify(publicKey, data, signature)
	assert.NoError(t, err)
	assert.True(t, valid)

	// Test with modified data
	modifiedData := []byte("Modified test message")
	valid, err = pqa.Verify(publicKey, modifiedData, signature)
	assert.NoError(t, err)
	assert.False(t, valid)
}

func TestEncryptAndDecrypt(t *testing.T) {
	pqa := NewPostQuantumAlgorithms()
	err := pqa.Initialize()
	assert.NoError(t, err)

	// Generate key pair
	publicKey, privateKey, err := pqa.GenerateKeyPair("CRYSTALS-Kyber")
	assert.NoError(t, err)

	// Test data
	data := []byte("Test message for encryption")

	// Encrypt data
	encrypted, err := pqa.Encrypt(publicKey, data)
	assert.NoError(t, err)
	assert.NotNil(t, encrypted)

	// Decrypt data
	decrypted, err := pqa.Decrypt(privateKey, encrypted)
	assert.NoError(t, err)
	assert.Equal(t, data, decrypted)
}

func TestGetAlgorithmSecurityLevel(t *testing.T) {
	pqa := NewPostQuantumAlgorithms()
	err := pqa.Initialize()
	assert.NoError(t, err)

	// Test with valid algorithm
	level, err := pqa.GetAlgorithmSecurityLevel("CRYSTALS-Kyber")
	assert.NoError(t, err)
	assert.Greater(t, level, 0)

	// Test with invalid algorithm
	_, err = pqa.GetAlgorithmSecurityLevel("InvalidAlgorithm")
	assert.Error(t, err)
}

func TestGetAlgorithmKeySize(t *testing.T) {
	pqa := NewPostQuantumAlgorithms()
	err := pqa.Initialize()
	assert.NoError(t, err)

	// Test with valid algorithm
	size, err := pqa.GetAlgorithmKeySize("CRYSTALS-Kyber")
	assert.NoError(t, err)
	assert.Greater(t, size, 0)

	// Test with invalid algorithm
	_, err = pqa.GetAlgorithmKeySize("InvalidAlgorithm")
	assert.Error(t, err)
}

func TestGetAlgorithmSignatureSize(t *testing.T) {
	pqa := NewPostQuantumAlgorithms()
	err := pqa.Initialize()
	assert.NoError(t, err)

	// Test with valid algorithm
	size, err := pqa.GetAlgorithmSignatureSize("CRYSTALS-Dilithium")
	assert.NoError(t, err)
	assert.Greater(t, size, 0)

	// Test with invalid algorithm
	_, err = pqa.GetAlgorithmSignatureSize("InvalidAlgorithm")
	assert.Error(t, err)
}

func TestGetDefaultAlgorithm(t *testing.T) {
	pqa := NewPostQuantumAlgorithms()
	err := pqa.Initialize()
	assert.NoError(t, err)

	// Get default algorithm
	algorithm, err := pqa.GetDefaultAlgorithm()
	assert.NoError(t, err)
	assert.NotEmpty(t, algorithm)

	// Verify it's a valid algorithm
	available, err := pqa.IsAlgorithmAvailable(algorithm)
	assert.NoError(t, err)
	assert.True(t, available)
}

func TestSetDefaultAlgorithm(t *testing.T) {
	pqa := NewPostQuantumAlgorithms()
	err := pqa.Initialize()
	assert.NoError(t, err)

	// Test with valid algorithm
	err = pqa.SetDefaultAlgorithm("CRYSTALS-Kyber")
	assert.NoError(t, err)

	// Verify the change
	algorithm, err := pqa.GetDefaultAlgorithm()
	assert.NoError(t, err)
	assert.Equal(t, "CRYSTALS-Kyber", algorithm)

	// Test with invalid algorithm
	err = pqa.SetDefaultAlgorithm("InvalidAlgorithm")
	assert.Error(t, err)
}

func TestOperationsAfterShutdown(t *testing.T) {
	pqa := NewPostQuantumAlgorithms()
	err := pqa.Initialize()
	assert.NoError(t, err)

	err = pqa.Shutdown()
	assert.NoError(t, err)

	// Test all operations after shutdown
	_, err = pqa.GetAvailableAlgorithms()
	assert.Error(t, err)

	_, err = pqa.GetAlgorithmInfo("CRYSTALS-Kyber")
	assert.Error(t, err)

	_, err = pqa.IsAlgorithmAvailable("CRYSTALS-Kyber")
	assert.Error(t, err)

	_, err = pqa.IsAlgorithmRecommended("CRYSTALS-Kyber")
	assert.Error(t, err)

	_, _, err = pqa.GenerateKeyPair("CRYSTALS-Kyber")
	assert.Error(t, err)

	_, err = pqa.GetDefaultAlgorithm()
	assert.Error(t, err)

	err = pqa.SetDefaultAlgorithm("CRYSTALS-Kyber")
	assert.Error(t, err)
}

File: bindings/typescript/tests/quantum/PostQuantumAlgorithms.test.ts
import { PostQuantumAlgorithms, AlgorithmInfo } from '../../src/quantum/PostQuantumAlgorithms';
import { SatoxError } from '../../src/types';

describe('PostQuantumAlgorithms', () => {
    let algorithms: PostQuantumAlgorithms;

    beforeEach(() => {
        algorithms = new PostQuantumAlgorithms();
        algorithms.initialize();
    });

    afterEach(() => {
        algorithms.shutdown();
    });

    describe('initialization', () => {
        it('should initialize and shutdown correctly', () => {
            const newAlgorithms = new PostQuantumAlgorithms();
            expect(newAlgorithms.isInitialized()).toBe(false);
            expect(newAlgorithms.initialize()).toBe(true);
            expect(newAlgorithms.isInitialized()).toBe(true);
            expect(newAlgorithms.shutdown()).toBe(true);
            expect(newAlgorithms.isInitialized()).toBe(false);
        });
    });

    describe('key pair generation', () => {
        it('should generate valid key pairs', async () => {
            const [publicKey, privateKey] = await algorithms.generateKeyPair('test-algo');
            expect(publicKey).toBeDefined();
            expect(privateKey).toBeDefined();
            expect(publicKey.length).toBeGreaterThan(0);
            expect(privateKey.length).toBeGreaterThan(0);
        });
    });

    describe('signing and verification', () => {
        it('should sign and verify data correctly', async () => {
            const [publicKey, privateKey] = await algorithms.generateKeyPair('test-algo');
            const data = new Uint8Array([1, 2, 3, 4, 5]);
            
            const signature = await algorithms.sign(privateKey, data);
            expect(signature).toBeDefined();
            expect(signature.length).toBeGreaterThan(0);
            
            expect(await algorithms.verify(publicKey, data, signature)).toBe(true);
            expect(await algorithms.verify(publicKey, data, new Uint8Array([0]))).toBe(false);
        });
    });

    describe('encryption and decryption', () => {
        it('should encrypt and decrypt data correctly', async () => {
            const [publicKey, privateKey] = await algorithms.generateKeyPair('test-algo');
            const data = new Uint8Array([1, 2, 3, 4, 5]);
            
            const encrypted = await algorithms.encrypt(publicKey, data);
            expect(encrypted).toBeDefined();
            expect(encrypted).not.toEqual(data);
            
            const decrypted = await algorithms.decrypt(privateKey, encrypted);
            expect(decrypted).toEqual(data);
        });
    });

    describe('algorithm information', () => {
        it('should get algorithm info', () => {
            const info = algorithms.getAlgorithmInfo('test-algo');
            expect(info).toBeDefined();
            expect(info.name).toBe('test-algo');
            expect(info.securityLevel).toBeGreaterThan(0);
            expect(info.keySize).toBeGreaterThan(0);
            expect(info.signatureSize).toBeGreaterThan(0);
            expect(typeof info.isRecommended).toBe('boolean');
            expect(typeof info.description).toBe('string');
        });

        it('should get available algorithms', () => {
            const available = algorithms.getAvailableAlgorithms();
            expect(available).toBeDefined();
            expect(Array.isArray(available)).toBe(true);
            expect(available.length).toBeGreaterThan(0);
        });

        it('should get recommended algorithms', () => {
            const recommended = algorithms.getRecommendedAlgorithms();
            expect(recommended).toBeDefined();
            expect(Array.isArray(recommended)).toBe(true);
            expect(recommended.length).toBeGreaterThan(0);
        });
    });

    describe('algorithm checks', () => {
        it('should check algorithm availability', () => {
            expect(algorithms.isAlgorithmAvailable('test-algo')).toBe(true);
            expect(algorithms.isAlgorithmAvailable('non-existent')).toBe(false);
        });

        it('should check algorithm recommendation', () => {
            const isRecommended = algorithms.isAlgorithmRecommended('test-algo');
            expect(typeof isRecommended).toBe('boolean');
        });
    });

    describe('algorithm properties', () => {
        it('should get security level', () => {
            const level = algorithms.getAlgorithmSecurityLevel('test-algo');
            expect(level).toBeGreaterThan(0);
        });

        it('should get key size', () => {
            const size = algorithms.getAlgorithmKeySize('test-algo');
            expect(size).toBeGreaterThan(0);
        });

        it('should get signature size', () => {
            const size = algorithms.getAlgorithmSignatureSize('test-algo');
            expect(size).toBeGreaterThan(0);
        });
    });

    describe('default algorithm', () => {
        it('should get and set default algorithm', () => {
            const defaultAlgo = algorithms.getDefaultAlgorithm();
            expect(defaultAlgo).toBeDefined();
            expect(typeof defaultAlgo).toBe('string');
            
            expect(algorithms.setDefaultAlgorithm('test-algo')).toBe(true);
            expect(algorithms.getDefaultAlgorithm()).toBe('test-algo');
        });
    });

    describe('error handling', () => {
        it('should throw when not initialized', async () => {
            const uninitializedAlgorithms = new PostQuantumAlgorithms();
            
            await expect(uninitializedAlgorithms.generateKeyPair('test-algo'))
                .rejects.toThrow(SatoxError);
            await expect(uninitializedAlgorithms.sign(new Uint8Array(), new Uint8Array()))
                .rejects.toThrow(SatoxError);
            await expect(uninitializedAlgorithms.verify(new Uint8Array(), new Uint8Array(), new Uint8Array()))
                .rejects.toThrow(SatoxError);
            await expect(uninitializedAlgorithms.encrypt(new Uint8Array(), new Uint8Array()))
                .rejects.toThrow(SatoxError);
            await expect(uninitializedAlgorithms.decrypt(new Uint8Array(), new Uint8Array()))
                .rejects.toThrow(SatoxError);
        });

        it('should throw on invalid inputs', async () => {
            await expect(algorithms.generateKeyPair(''))
                .rejects.toThrow(SatoxError);
            await expect(algorithms.sign(null as any, new Uint8Array()))
                .rejects.toThrow(SatoxError);
            await expect(algorithms.verify(null as any, new Uint8Array(), new Uint8Array()))
                .rejects.toThrow(SatoxError);
            await expect(algorithms.encrypt(null as any, new Uint8Array()))
                .rejects.toThrow(SatoxError);
            await expect(algorithms.decrypt(null as any, new Uint8Array()))
                .rejects.toThrow(SatoxError);
        });
    });

    describe('algorithm consistency', () => {
        it('should maintain consistency across operations', async () => {
            const [publicKey, privateKey] = await algorithms.generateKeyPair('test-algo');
            const data = new Uint8Array([1, 2, 3, 4, 5]);
            
            // Test encryption/decryption consistency
            const encrypted = await algorithms.encrypt(publicKey, data);
            const decrypted = await algorithms.decrypt(privateKey, encrypted);
            expect(decrypted).toEqual(data);
            
            // Test signing/verification consistency
            const signature = await algorithms.sign(privateKey, data);
            expect(await algorithms.verify(publicKey, data, signature)).toBe(true);
        });
    });
}); 